flatten_json.go

```diff
@@ -208,42 +227,56 @@ func (fj *flattenJSON) readObject(pathName []byte) error {
 				val, alt, err = fj.readNumber()
 				isLeaf = true
 			case '[':
-				if !fj.tracker.IsNameUsed(memberName) {
+				if !pathNode.IsNameUsed(memberName) {
 					fj.skipping++
 				}
-				var arrayPath []byte
-				if fj.skipping == 0 {
-					arrayPath = pathForChild(pathName, memberName)
-				}
 
 				if fj.skipping > 0 || !memberIsUsed {
 					err = fj.skipBlock('[', ']')
 				} else {
-					err = fj.readArray(arrayPath)
+					arrayPathNode, exists := pathNode.Get(memberName)
+					if !exists {
+						// Arrays are interesting, they can be field or node.
+						// Given this case:
+						//  { "geo": { "coords": [{"coordinates": [1,2,3]}] } }
+						// "coords" is a node.
+						// "coordinates" is a field.
+						arrayPathNode = pathNode
+					}
+
+					err = fj.readArray(pathNode.PathForMemberName(memberName), arrayPathNode)
 				}
 				if err != nil {
 					return err
 				}
-				if !fj.tracker.IsNameUsed(memberName) {
+				if !pathNode.IsNameUsed(memberName) {
 					fj.skipping--
 				}
 			case '{':
-				if !fj.tracker.IsNameUsed(memberName) {
+				if !pathNode.IsNameUsed(memberName) {
 					fj.skipping++
 				}
-				var objectPath []byte
-				if fj.skipping == 0 {
-					objectPath = pathForChild(pathName, memberName)
-				}
 				if fj.skipping > 0 || !memberIsUsed {
 					err = fj.skipBlock('{', '}')
 				} else {
-					err = fj.readObject(objectPath)
+					objectPathNode, exists := pathNode.Get(memberName)
+					if !exists {
+						// Interesting case.
+						// Looks like it's an object, which not registered as object (node) in our paths index.
+						// Which means someone is doing matching on object field as primitive field, for example:
+						//  pattern: `{ "a": [1] }`, event: `{ "a": { "b": 1 } }` for example.
+						return fj.error(fmt.Sprintf("[%s] doesn't exists in path index, but we still got here [%s].\n", string(memberName), pathNode.String()))
```

Comment:
Haven't fully studied yet, thinking out loud: Is this a situation that can arise, i.e. if you add that pattern and match that event, would you get to this `fj.error`?  If so, it shouldn't be an error, it should be just a no-match return?  If you can't get to this error, should the code even exist?

Reply:
Ignore the error returning here - I did this for debug purposes and in the final PR it should be removed.

I wasn't sure (and didn't checked) if we support select objects as `quamina.Field`, the use-case I have in mind is doing `exists: true` on object (this is something I plan doing).

An example for that will be, given this event: `{ "name": "yosi", "changes": { "name": ["yosy", "yosi"] } }` I would like to get match only when `changes` exists in the object (since some events might not contains `changes` json property).

If we want to support in the future/currently support it, it means that this code should get same handling I did for arrays: https://github.com/timbray/quamina/pull/124/files#diff-745034a9cc805c7b30a993232b1a6d53132a7e2099d5b72641b40ddfa21e3916R239

Reply:
Right, at the moment `"exists":true` only works for leaf nodes, because of the way flattening works. I would 100% agree in principle with fixing this but it would be a nontrivial change to the state-machine model.  I suggest that we shouldn't hold up this PR for that enhancement but let's record an issue so we don't lose the idea.

Reply:
I checked again for what I specifically need, and I was wrong:
```
{ "name": "yosi", "changes": { "name": ["yosy", "yosi"] } }
```

I would need `exists: true` on `changes.name` and not on `changes`.

Now the question is:
* Do we still want an issue for allowing `exists: true` on object nodes?
* How should I handle this case, because currently it's an error, I will try to reproduce in main and see what's the behaviour.

Reply:
Fixed the login here, and added tests for this to match the main branch:
https://github.com/timbray/quamina/pull/124/commits/1080360922fe4a84208fb1c5070a9f35eeee7b6a

Once we want to add support for "exists" on objects we can revisit this.